Update upsplus.py

deHarro opened the pull request:
I have an UPSplus which, from time to time, stops charging, even the charger is still delivering power to the system.
In this situation the Raspberry is pwoered fro mthe batteries and they are getting empty. With the current implementation the graceful shutdown is never initialized since this case lies in the else clauss.
By moving the battery interrogation one indentation outwards, this erronous case is caught sanely.
Harry

hellresistor commented:
@deHarro I am getting that issue. I will wait about pull request aprovement.

deHarro commented:
Wow!
I thougt I am the only one with this problem.
I recently asked the developers what mechanism on the UPS causes this shutting down of the chargers power path.

You know, my proposal is only a workaround for preventing the RPi from destroaing its SD card ;-)

I think there is a firmware glitch which causes this behavior of the UPS. And glitches are very difficult to find :(

Harry

hellresistor commented:
I unfourtunly unable help on this :S was not my 'thing' ^^ 
just only with bash scripting

nickfox-taterli commented:
Every time I resample, it starts charging again, did you check this setting?

[How_To_Adjust_The_Sampling_Period](https://wiki.52pi.com/index.php/UPS_Plus_SKU:_EP-0136#How_To_Adjust_The_Sampling_Period)

When you remove the power supply, the battery will start to discharge. When the battery is already discharging before you remove the power supply, it only means that the input power is less than the power required by the Raspberry Pi.

Instead, when you don't remove the power supply, there are two power supply paths that prioritise keeping the Raspberry Pi powered, and secondly will power the battery, which will be re-sampled after the [Sampling Period] when it is full, and will be re-charged if the capacity drops to about 90%.

deHarro commented:
Hi @nickfox-taterli.

yes, I read, checked and understand that sampling thing.
According to the advices of your after-sales staff, I changed the sampling period to be 1 minute (from default 2 minutes).

The main thing in this discussion is, **I do not** remove the charger.
The charer is all the time attached to the UPS and delivers power to the UPS.

Despite this fact, suddenly the UPS stops powering the system ("system" is UPS + RPi) from the charger and switches to battery operation. 
As you can see from my traces - I sent them to the after-sales staff already - the voltage on UPS input is there, but the UPS powers the RPi from battery.

I have a commented short form of a trace:
```
2021-05-12 - 11:57:02 -> Vbat 4,208 V, Ibat   -38,0 mA, Pbat   48,8 mW - Vrpi 4,948 V, Irpi  3180,0 mA, Prpi 15829,3 mW, - Vusb 5,134 V <<-- start of trace
2021-05-12 - 11:57:05 -> Vbat 4,208 V, Ibat   -46,1 mA, Pbat   97,6 mW - Vrpi 4,956 V, Irpi  3162,7 mA, Prpi 15630,7 mW, - Vusb 5,155 V
2021-05-12 - 11:57:08 -> Vbat 4,212 V, Ibat   -18,0 mA, Pbat  200,0 mW - Vrpi 4,944 V, Irpi  3145,6 mA, Prpi 15613,2 mW, - Vusb 5,184 V
2021-05-12 - 11:57:11 -> Vbat 4,208 V, Ibat   -32,0 mA, Pbat   63,4 mW - Vrpi 4,952 V, Irpi  3147,0 mA, Prpi 15602,8 mW, - Vusb 5,197 V
2021-05-12 - 11:57:15 -> Vbat 4,208 V, Ibat   -53,9 mA, Pbat  161,0 mW - Vrpi 4,988 V, Irpi  3124,2 mA, Prpi 15613,2 mW, - Vusb 5,184 V
2021-05-12 - 11:57:18 -> Vbat 4,208 V, Ibat    -2,0 mA, Pbat  165,9 mW - Vrpi 4,944 V, Irpi  3147,0 mA, Prpi 15592,3 mW, - Vusb 5,175 V
2021-05-12 - 11:57:21 -> Vbat 4,204 V, Ibat   -16,1 mA, Pbat   53,7 mW - Vrpi 4,984 V, Irpi  3138,5 mA, Prpi 15613,2 mW, - Vusb 5,194 V
2021-05-12 - 11:57:24 -> Vbat 4,204 V, Ibat   -16,1 mA, Pbat  200,0 mW - Vrpi 4,956 V, Irpi  3169,9 mA, Prpi 15700,3 mW, - Vusb 5,162 V
2021-05-12 - 11:57:27 -> Vbat 4,212 V, Ibat   -16,1 mA, Pbat   48,8 mW - Vrpi 4,984 V, Irpi  3147,0 mA, Prpi 15700,3 mW, - Vusb 5,207 V
2021-05-12 - 11:57:31 -> Vbat 4,208 V, Ibat    10,0 mA, Pbat   48,8 mW - Vrpi 4,952 V, Irpi  3147,0 mA, Prpi 15561,0 mW, - Vusb 5,162 V
2021-05-12 - 11:57:34 -> Vbat 4,208 V, Ibat   -18,0 mA, Pbat  239,0 mW - Vrpi 4,980 V, Irpi  3134,1 mA, Prpi 15613,2 mW, - Vusb 5,165 V
2021-05-12 - 11:57:37 -> Vbat 4,212 V, Ibat     3,9 mA, Pbat   24,4 mW - Vrpi 4,988 V, Irpi  3139,9 mA, Prpi 15700,3 mW, - Vusb 5,162 V
2021-05-12 - 11:57:40 -> Vbat 4,212 V, Ibat    -3,9 mA, Pbat  112,2 mW - Vrpi 4,956 V, Irpi  3155,6 mA, Prpi 15578,4 mW, - Vusb 5,165 V
2021-05-12 - 11:57:43 -> Vbat 4,208 V, Ibat   -18,0 mA, Pbat   87,8 mW - Vrpi 4,952 V, Irpi  3111,3 mA, Prpi 15466,9 mW, - Vusb 5,116 V
2021-05-12 - 11:57:47 -> Vbat 4,204 V, Ibat   -12,0 mA, Pbat  141,5 mW - Vrpi 4,968 V, Irpi  3111,3 mA, Prpi 15494,8 mW, - Vusb 5,116 V
2021-05-12 - 11:57:50 -> Vbat 4,208 V, Ibat   -36,1 mA, Pbat   53,7 mW - Vrpi 4,984 V, Irpi  3132,8 mA, Prpi 15693,4 mW, - Vusb 5,171 V
2021-05-12 - 11:57:53 -> Vbat 4,204 V, Ibat     0,0 mA, Pbat   97,6 mW - Vrpi 4,980 V, Irpi  3132,8 mA, Prpi 15613,2 mW, - Vusb 5,116 V
2021-05-12 - 11:57:56 -> Vbat 4,208 V, Ibat   -18,0 mA, Pbat   97,6 mW - Vrpi 4,976 V, Irpi  3101,4 mA, Prpi 15494,8 mW, - Vusb 5,155 V
2021-05-12 - 11:57:59 -> Vbat 4,208 V, Ibat    -3,9 mA, Pbat   39,0 mW - Vrpi 4,972 V, Irpi  3141,3 mA, Prpi 15613,2 mW, - Vusb 5,110 V
2021-05-12 - 11:58:03 -> Vbat 4,208 V, Ibat   -33,9 mA, Pbat  224,4 mW - Vrpi 4,952 V, Irpi  3152,8 mA, Prpi 15641,1 mW, - Vusb 5,216 V
2021-05-12 - 11:58:06 -> Vbat 4,204 V, Ibat   -16,1 mA, Pbat   97,6 mW - Vrpi 4,972 V, Irpi  3117,1 mA, Prpi 15487,8 mW, - Vusb 5,155 V
2021-05-12 - 11:58:09 -> Vbat 4,208 V, Ibat     2,0 mA, Pbat   34,1 mW - Vrpi 4,948 V, Irpi  3158,5 mA, Prpi 15648,1 mW, - Vusb 5,175 V
2021-05-12 - 11:58:12 -> Vbat 4,208 V, Ibat   -28,0 mA, Pbat   48,8 mW - Vrpi 4,952 V, Irpi  3142,7 mA, Prpi 15620,2 mW, - Vusb 5,091 V
2021-05-12 - 11:58:15 -> Vbat 4,212 V, Ibat    26,1 mA, Pbat   97,6 mW - Vrpi 4,984 V, Irpi  3142,7 mA, Prpi 15682,9 mW, - Vusb 5,155 V <<-- time gap 1 start
...
2021-05-13 - 21:09:08 -> Vbat 4,200 V, Ibat   -22,0 mA, Pbat   209,8 mW - Vrpi 4,984 V, Irpi  3098,4 mA, Prpi 15540,1 mW, - Vusb 5,177 V <<-- time gap 1 end
2021-05-13 - 21:09:11 -> Vbat 4,212 V, Ibat    -6,1 mA, Pbat     9,8 mW - Vrpi 4,988 V, Irpi  3095,6 mA, Prpi 15386,8 mW, - Vusb 5,125 V
2021-05-13 - 21:09:14 -> Vbat 4,204 V, Ibat    -8,0 mA, Pbat     9,8 mW - Vrpi 4,992 V, Irpi  3098,4 mA, Prpi 15512,2 mW, - Vusb 5,174 V
2021-05-13 - 21:09:18 -> Vbat 4,204 V, Ibat     2,0 mA, Pbat    24,4 mW - Vrpi 4,980 V, Irpi  3142,7 mA, Prpi 15686,4 mW, - Vusb 5,145 V
2021-05-13 - 21:09:21 -> Vbat 4,208 V, Ibat    -3,9 mA, Pbat    53,7 mW - Vrpi 4,984 V, Irpi  3124,2 mA, Prpi 15630,7 mW, - Vusb 5,125 V
2021-05-13 - 21:09:24 -> Vbat 4,200 V, Ibat   -22,0 mA, Pbat   165,9 mW - Vrpi 4,988 V, Irpi  3077,0 mA, Prpi 15466,9 mW, - Vusb 5,174 V
2021-05-13 - 21:09:27 -> Vbat 4,204 V, Ibat   -16,1 mA, Pbat   200,0 mW - Vrpi 4,984 V, Irpi  3077,0 mA, Prpi 15324,0 mW, - Vusb 5,157 V
2021-05-13 - 21:09:30 -> Vbat 4,204 V, Ibat     2,0 mA, Pbat   112,2 mW - Vrpi 4,948 V, Irpi  3180,0 mA, Prpi 15773,5 mW, - Vusb 5,174 V
2021-05-13 - 21:09:34 -> Vbat 4,204 V, Ibat   -40,0 mA, Pbat    63,4 mW - Vrpi 4,968 V, Irpi  3141,3 mA, Prpi 15703,8 mW, - Vusb 5,145 V
2021-05-13 - 21:09:37 -> Vbat 4,204 V, Ibat    -8,0 mA, Pbat    73,2 mW - Vrpi 4,980 V, Irpi  3088,5 mA, Prpi 15327,5 mW, - Vusb 5,138 V
2021-05-13 - 21:09:40 -> Vbat 4,204 V, Ibat   -12,0 mA, Pbat    53,7 mW - Vrpi 4,960 V, Irpi  3130,0 mA, Prpi 15557,5 mW, - Vusb 5,141 V
2021-05-13 - 21:09:43 -> Vbat 4,204 V, Ibat     0,0 mA, Pbat   136,6 mW - Vrpi 4,980 V, Irpi  3094,3 mA, Prpi 15397,2 mW, - Vusb 5,153 V
2021-05-13 - 21:09:46 -> Vbat 4,208 V, Ibat   -23,9 mA, Pbat    97,6 mW - Vrpi 4,976 V, Irpi  3087,1 mA, Prpi 15324,0 mW, - Vusb 5,153 V
2021-05-13 - 21:09:50 -> Vbat 4,204 V, Ibat   -10,0 mA, Pbat    63,4 mW - Vrpi 4,980 V, Irpi  3088,5 mA, Prpi 15449,5 mW, - Vusb 5,146 V
2021-05-13 - 21:09:53 -> Vbat 4,172 V, Ibat -1932,0 mA, Pbat  8034,1 mW - Vrpi 4,960 V, Irpi  3130,0 mA, Prpi 15522,6 mW, - Vusb 5,230 V <<-- charger still connected, UPS switched to battery operation
2021-05-13 - 21:09:56 -> Vbat 4,176 V, Ibat -1842,0 mA, Pbat  7814,6 mW - Vrpi 4,988 V, Irpi  3111,3 mA, Prpi 15522,6 mW, - Vusb 5,252 V
2021-05-13 - 21:09:59 -> Vbat 4,168 V, Ibat -1980,0 mA, Pbat  8209,8 mW - Vrpi 4,960 V, Irpi  3119,9 mA, Prpi 15512,2 mW, - Vusb 5,220 V
2021-05-13 - 21:10:02 -> Vbat 4,172 V, Ibat -1905,9 mA, Pbat  7912,2 mW - Vrpi 4,964 V, Irpi  3130,0 mA, Prpi 15522,6 mW, - Vusb 5,249 V
2021-05-13 - 21:10:06 -> Vbat 4,168 V, Ibat -1992,0 mA, Pbat  8273,2 mW - Vrpi 4,976 V, Irpi  3132,8 mA, Prpi 15641,1 mW, - Vusb 5,226 V
2021-05-13 - 21:10:09 -> Vbat 4,168 V, Ibat -1932,0 mA, Pbat  8224,4 mW - Vrpi 4,960 V, Irpi  3135,5 mA, Prpi 15595,8 mW, - Vusb 5,207 V
2021-05-13 - 21:10:12 -> Vbat 4,172 V, Ibat -1833,9 mA, Pbat  7595,1 mW - Vrpi 4,972 V, Irpi  3064,1 mA, Prpi 15351,9 mW, - Vusb 5,249 V
2021-05-13 - 21:10:15 -> Vbat 4,172 V, Ibat -1850,0 mA, Pbat  7775,6 mW - Vrpi 4,976 V, Irpi  3082,8 mA, Prpi 15299,7 mW, - Vusb 5,230 V
2021-05-13 - 21:10:18 -> Vbat 4,164 V, Ibat -1973,9 mA, Pbat  7673,2 mW - Vrpi 4,980 V, Irpi  3085,5 mA, Prpi 15282,2 mW, - Vusb 5,246 V
2021-05-13 - 21:10:22 -> Vbat 4,164 V, Ibat -2102,0 mA, Pbat  8736,6 mW - Vrpi 4,944 V, Irpi  3180,0 mA, Prpi 15738,7 mW, - Vusb 5,220 V
2021-05-13 - 21:10:25 -> Vbat 4,168 V, Ibat -1912,0 mA, Pbat  7922,0 mW - Vrpi 4,976 V, Irpi  3130,0 mA, Prpi 15738,7 mW, - Vusb 5,223 V
2021-05-13 - 21:10:28 -> Vbat 4,168 V, Ibat -1800,0 mA, Pbat  7502,4 mW - Vrpi 4,984 V, Irpi  3132,8 mA, Prpi 15595,8 mW, - Vusb 5,223 V
2021-05-13 - 21:10:31 -> Vbat 4,168 V, Ibat -1865,9 mA, Pbat  7736,6 mW - Vrpi 4,972 V, Irpi  3088,5 mA, Prpi 15372,8 mW, - Vusb 5,162 V
2021-05-13 - 21:10:34 -> Vbat 4,164 V, Ibat -1873,9 mA, Pbat  7746,3 mW - Vrpi 4,948 V, Irpi  3171,3 mA, Prpi 15703,8 mW, - Vusb 5,249 V <<-- time gap 2 start
...
2021-05-13 - 22:48:16 -> Vbat 3,652 V, Ibat -2243,9 mA, Pbat  8185,4 mW - Vrpi 4,952 V, Irpi  3144,3 mA, Prpi 15515,7 mW, - Vusb 5,239 V <<-- time gap 2 end
2021-05-13 - 22:48:19 -> Vbat 3,652 V, Ibat -2150,0 mA, Pbat  7770,7 mW - Vrpi 4,952 V, Irpi  3122,8 mA, Prpi 15351,9 mW, - Vusb 5,261 V
2021-05-13 - 22:48:22 -> Vbat 3,656 V, Ibat -2247,8 mA, Pbat  8273,2 mW - Vrpi 4,948 V, Irpi  3142,7 mA, Prpi 15557,5 mW, - Vusb 5,239 V
2021-05-13 - 22:48:25 -> Vbat 3,660 V, Ibat -2040,0 mA, Pbat  7482,9 mW - Vrpi 4,948 V, Irpi  3125,6 mA, Prpi 15470,4 mW, - Vusb 5,213 V
2021-05-13 - 22:48:28 -> Vbat 3,652 V, Ibat -2025,9 mA, Pbat  7409,8 mW - Vrpi 4,956 V, Irpi  3101,4 mA, Prpi 15477,4 mW, - Vusb 5,255 V
2021-05-13 - 22:48:32 -> Vbat 3,656 V, Ibat -2127,8 mA, Pbat  7902,4 mW - Vrpi 4,956 V, Irpi  3124,2 mA, Prpi 15567,9 mW, - Vusb 5,265 V
2021-05-13 - 22:48:35 -> Vbat 3,660 V, Ibat -2133,9 mA, Pbat  7697,6 mW - Vrpi 4,952 V, Irpi  3124,2 mA, Prpi 15550,5 mW, - Vusb 5,242 V
2021-05-13 - 22:48:38 -> Vbat 3,652 V, Ibat -2223,9 mA, Pbat  8151,2 mW - Vrpi 4,948 V, Irpi  3157,0 mA, Prpi 15658,5 mW, - Vusb 5,261 V
2021-05-13 - 22:48:41 -> Vbat 3,656 V, Ibat -2100,0 mA, Pbat  7673,2 mW - Vrpi 4,948 V, Irpi  3119,9 mA, Prpi 15470,4 mW, - Vusb 5,213 V
2021-05-13 - 22:48:44 -> Vbat 3,660 V, Ibat -2125,9 mA, Pbat  7897,6 mW - Vrpi 4,948 V, Irpi  3101,4 mA, Prpi 15344,9 mW, - Vusb 5,226 V
2021-05-13 - 22:48:48 -> Vbat 3,664 V, Ibat -2130,0 mA, Pbat  7878,0 mW - Vrpi 4,952 V, Irpi  3112,7 mA, Prpi 15407,7 mW, - Vusb 5,242 V
2021-05-13 - 22:48:51 -> Vbat 3,652 V, Ibat -2263,9 mA, Pbat  8409,8 mW - Vrpi 4,948 V, Irpi  3162,7 mA, Prpi 15669,0 mW, - Vusb 5,261 V
2021-05-13 - 22:48:54 -> Vbat 3,656 V, Ibat -2040,0 mA, Pbat  7434,1 mW - Vrpi 4,948 V, Irpi  3105,6 mA, Prpi 15355,4 mW, - Vusb 5,255 V
2021-05-13 - 22:48:57 -> Vbat 3,652 V, Ibat -2220,0 mA, Pbat  8097,6 mW - Vrpi 4,948 V, Irpi  3134,1 mA, Prpi 15515,7 mW, - Vusb 5,255 V
2021-05-13 - 22:49:00 -> Vbat 3,656 V, Ibat -2282,0 mA, Pbat  8278,0 mW - Vrpi 4,948 V, Irpi  3148,4 mA, Prpi 15578,4 mW, - Vusb 5,226 V
2021-05-13 - 22:49:04 -> Vbat 3,656 V, Ibat -2285,9 mA, Pbat  8473,2 mW - Vrpi 4,956 V, Irpi  3147,0 mA, Prpi 15585,4 mW, - Vusb 5,219 V
2021-05-13 - 22:49:07 -> Vbat 3,656 V, Ibat -2275,9 mA, Pbat  8248,8 mW - Vrpi 4,948 V, Irpi  3145,6 mA, Prpi 15529,6 mW, - Vusb 5,245 V
2021-05-13 - 22:49:10 -> Vbat 3,652 V, Ibat -2063,9 mA, Pbat  7482,9 mW - Vrpi 4,952 V, Irpi  3095,6 mA, Prpi 15338,0 mW, - Vusb 5,219 V
2021-05-13 - 22:49:13 -> Vbat 3,652 V, Ibat -2133,9 mA, Pbat  7897,6 mW - Vrpi 4,948 V, Irpi  3088,5 mA, Prpi 15355,4 mW, - Vusb 5,232 V
2021-05-13 - 22:49:16 -> Vbat 3,652 V, Ibat -2097,8 mA, Pbat  7770,7 mW - Vrpi 4,956 V, Irpi  3112,7 mA, Prpi 15344,9 mW, - Vusb 5,226 V
2021-05-13 - 22:49:20 -> Vbat 3,656 V, Ibat -2113,9 mA, Pbat  7863,4 mW - Vrpi 4,952 V, Irpi  3099,8 mA, Prpi 15338,0 mW, - Vusb 5,245 V
2021-05-13 - 22:49:23 -> Vbat 3,652 V, Ibat -2247,8 mA, Pbat  8097,6 mW - Vrpi 4,944 V, Irpi  3145,6 mA, Prpi 15561,0 mW, - Vusb 5,274 V <<-- end of trace
--- RPi powered down and switched OFF

```
The values for current (Irpi) and power (Prpi) for the RPi () are 5.5 times too high, since I didn't change the default settings in the Python script. This I did afterwards (the trace is from an early stage of testing).

As test I inserted a USB power gauge into the USB line from the charger and see, that in the error case (after several hours), there is no (zero) current flowing from the charger to the UPS, but voltage is at ~5.1 V, still.

As far as I see it, the µC of the UPS seems to switch off the power path from the charger to the UPS or at least from the charger to the RPi. 
This behavior is definitely to be named an error. 
Without any knowledge of the circuitry of the UPS I cannot support more than that.

If you have any idea what I can do as test, please let me know.

Harry

hellresistor commented:
@deHarro Heya, i have a question , this its hardware problem? or just firmware?

hellresistor commented:
@deHarro
A curiosity. how much amp  delivery your power source on output? (3Amp)

deHarro commented:
@hellresistor 
My charger delivers 3A @ 5,1 V. It is a genuine Raspberry Pi 4 charger with USB C plug.

Harry

hellresistor commented:
Add more info about my configs. I am using the RTC Clock config. `addr=0x68` on dtparams. Tomorrow will try with just basic configuration.

deHarro commented:
Just a note on closing this request:
As you can see (by checking my changes to your script), the changes do not alter the behaviour of the script under "normal" conditions.
Only if the UPS runs into the error condition I informed you about, then the changes allow the script to shut down the Raspberry Pi when the batteries are drained.
So my changes are saving the batteries from beeing deep discharged.

**Have in mind, that deep discharging LiIon or LiPo batteries may result in severe and hazardous situations.**

**So I appeal to you, just for your own sake, to implement my changes anyway, or any other changes to prevent deep discharging the batteries in an error condition of the UPS hardware or software/firmware.
This is general practice and should be observed always!**

Harry
